src/xdsl/parser.py

```diff
@@ -244,7 +244,7 @@ def parse_optional_int_literal(self,
         res = self.parse_while(lambda char: char.isnumeric(),
                                skip_white_space=False)
         if len(res) == 0:
-            if is_negative is not None:
+            if is_negative:
```

Comment:
Here, this works only because `is_negative` is either `True` or `None`. We should have a comment here specifying this.

Reply:
However, it would be better to return `True` or `False`, right?

Reply:
So the reason why I'm always returning `None` is because I have some "combinator" methods that expects the parser to return `None` if nothing was parsed.
Essentially, each parsing function is returning an `Optional` of the thing we are parsing. So parsing an expected string has really only two states, `True` and `None`. Though if you want we could return the parsed string, but since it should be equal to the one given, I didn't felt it was a good design.

Reply:
I am of the opinion that everything that `may not work` should be accompanied by a test to prove that.
Since we know that its either True or None, ok....but the point of having sth boolean like True...is to also have something False at any point. I would vote for either True/False to string/None.
Thoughts?

Reply:
I second @georgebisbas's opinion. I personally feel that True/False is closer to the current approach, so the step might be smaller, but I am fine with both.

I also feel that this should not be part of this PR as it seems to be a quite large refactoring but rather its own one.

Reply:
I would prefer returning `string/None` then :/
So what I wanted to have with this parser is a "unified" way of writing parsing rules, so we can use those combinators.
That's why I would prefer returning `string/None`, since returning `False` would mean that you are parsing something that has a `False` meaning, rather than `None` which means parsing nothing.

Reply:
For instance, for `parse_list`, we expect a parsing function, that parses optionally something. Currently, the rule is that if the labmda is returning `None`, then it means it couldn't parse the next element.
Switching to `True/False` would mean that we wouldn't be able to use this function in `parse_list`

Reply:
I don't think I quite understand. You are saying that the behavior when parsing a string like "False" would be weird?

Why wouldn't we be able to use the function, if we changed the semantics to "if the lambda is returning False, it means it couldn't parse the next element"?

What do you mean with unified way? What different kinds of parsing functions are there? Do you mean the ones with try and the ones without?

Also, do we want to move this discussion to an issue? I feel that would make this easier to find later.

Reply:
Has there been any discussion on this?